src/single_file/c/illegal_instruction_experiments.c

```diff
@@ -0,0 +1,82 @@
+/**
+ * @file    illegal_instruction_experiments.c
+ * @brief   Create illegal instruction exceptions for development of emulated instructions in ogsbi.
+ * 
+ * @copyright
+ *  Copyright (C) 2024 Eric Jessee\n
+ *  See the LICENSE file at the root of the project for licensing info.
+*/
+
+/* ------------------------------------------------------------------------------------------------
+ * Includes
+ * --------------------------------------------------------------------------------------------- */
+
+#include <assert.h>
+#include <stdio.h>
+#include <stdbool.h>
+#include <stdlib.h>
+#include <stdint.h>
+
+/* ------------------------------------------------------------------------------------------------
+ * Function Implementations
+ * --------------------------------------------------------------------------------------------- */
+
+int main(void) {
+    puts("forcing a mulh instruction");
+    __asm__ volatile (
+        "li a7, -100\n"
+        "li a6, 10\n"
+        "mulh a5, a6, a7"
```

Comment:
Good using `volatile` here but there is a slight problem: you're overwriting `a7` and `a6` which the compiler can't predict and may clobber important values. Additionally it would be nice to have an `assert` here to automatically detect later breakage.

GCC has some nice extended syntax that allows you to access C variables from assembly! This prevents you from clobbering anything the compiler didn't expect and lets you use the values later (ex. in an assert)!

You probably want something like this (untested):
```
uint32_t a = 10;
uint32_t b = -100;
uint32_t c;
__asm__ volatile (
    "mulh %[dest], %[left_op], %[right_op]"
    : [dest] "=r" (c)
    : [left_op] "r" (a), [right_op] "r" (b)
    : /* No additional clobbered registers */
);
assert(c == 0xFFFFFFFF);
```

(The other assembly patches below need similar tweaks).

Reply:
Yep, totally makes sense. I've seen these techniques elsewhere in rvsw but I was struggling to understand the syntax. This helps though- will add these tweaks as suggested